.github/workflows/validate.yml

```diff
@@ -119,222 +128,158 @@ jobs:
-          cabal-version: 3.12.1.0 # see https://github.com/haskell/cabal/pull/10251
-          ghcup-release-channel: https://raw.githubusercontent.com/haskell/ghcup-metadata/master/ghcup-prereleases-0.0.8.yaml
+          shell: ${{ matrix.sys.shell }}
+          ghc: ${{ matrix.ghc }}
+          allow-newer: ${{ github.event_name == 'workflow_dispatch' && github.event.inputs.allow_newer || '' }}
+          constraints: ${{ github.event_name == 'workflow_dispatch' && github.event.inputs.constraints || '' }}
 
-      #  See the following link for a breakdown of the following step
-      #  https://github.com/haskell/actions/issues/7#issuecomment-745697160
-      - uses: actions/cache@v4
-        with:
-          # validate.sh uses a special build dir
-          path: |
-            ${{ steps.setup-haskell.outputs.cabal-store }}
-            dist-*
-          key: ${{ runner.os }}-${{ matrix.ghc }}-${{ github.sha }}
-          restore-keys: ${{ runner.os }}-${{ matrix.ghc }}-
-
-      # The tool is not essential to the rest of the test suite. If
-      # hackage-repo-tool is not present, any test that requires it will
-      # be skipped.
-      # We want to keep this in the loop but we don't want to fail if
-      # hackage-repo-tool breaks or fails to support a newer GHC version.
-      - name: Install hackage-repo-tool
-        continue-on-error: true
-        run: cabal install --ignore-project hackage-repo-tool
-
-      # Needed by cabal-testsuite/PackageTests/Configure/setup.test.hs
-      - name: "MAC: Install Autotools"
-        if: runner.os == 'macOS'
-        run: brew install automake
-
-      # Needed by cabal-testsuite/PackageTests/Configure/setup.test.hs
-      - name: "WIN: Install Autotools"
-        if: runner.os == 'Windows'
-        run: /usr/bin/pacman --noconfirm -S autotools
-
-      - name: Set validate inputs
-        run: |
-          FLAGS="${{ env.COMMON_FLAGS }}"
-          if [[ "${{ matrix.ghc }}" == "${{ env.GHC_FOR_SOLVER_BENCHMARKS }}" ]]; then
-            FLAGS="$FLAGS --solver-benchmarks"
-          fi
-          if [[ "${{ matrix.ghc }}" == "${{ env.GHC_FOR_COMPLETE_HACKAGE_TESTS }}" ]]; then
-            FLAGS="$FLAGS --complete-hackage-tests"
-          fi
-          echo "FLAGS=$FLAGS" >> "$GITHUB_ENV"
-
-      - name: Validate print-config
-        run: sh validate.sh $FLAGS -s print-config
-
-      - name: Validate print-tool-versions
-        run: sh validate.sh $FLAGS -s print-tool-versions
-
-      - name: Validate build
-        run: sh validate.sh $FLAGS -s build
+  validate-tests:
+    name: Validate ${{ matrix.test }} ${{ matrix.sys.os }} ghc-${{ matrix.ghc }}
+    runs-on: ${{ matrix.sys.os }}
+    needs: [config, validate-build]
+    strategy:
+      fail-fast: false
+      # this matrix must match the one in `validate-build` (no, there's no way to share
+      # them, naturally) aside from the addition of `test` here
+      matrix:
+        test: [lib-tests, lib-suite, cli-tests, cli-suite]
```

Comment:
To make the list of jobs more manageable without losing (much of) end-to-end time benefits, we could have just too: `all-but-cli-suite` and `cli-suite`. I didn't check that it doesn't lose anything but I expect it not to.

Reply:
Unfortunately, that's much harder to wedge into this. If I split all the tests, it's a 3-line addition to the strategy matrix; if I split just one, I need to duplicate the whole job.

Reply:
(This is the stupidity of YAML-as-code again.)

Reply:
I'm very sorry but I don't understand. The test parameter of the matrix seems to be only used in

```
validate.sh $FLAGS -s ${{ matrix.test }}
```
So, if you teach validate.sh about the meta target `all-tests-but-cli-suite`, it should be trivial to use the matrix with two test values?

Reply:
Without touching validate.sh, you could check the test before right in the step that calls validate.sh, and on the meta make do a little bash loop.

Reply:
It's `cabal-validate` these days, and I think it's a bit more complicated than that at this point. And as to the bash loop, I thought you disapproved of those after the first one?

Reply:
I'm not a fan of those but 100+ jobs sounds even less sexy...